articles/node_lifecycle.md

```diff
@@ -0,0 +1,287 @@
+  in ROS 2. It has been written with consideration for the existing design of
+  the ROS 2 C++ client library, and in particular the current design of
+  executors.
+author: '[Geoffrey Biggs](https://github.com/gbiggs)'
+published: true
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Background
+
+A managed life cycle for nodes allows greater precision over the state of ROS
+system. It will allow roslaunch to have greater control over launching a ROS
+system, and in particular will allow roslaunch to ensure that all components
+have been instantiated correctly before it allows any component to begin
+executing its behaviour. It will also allow nodes to be restarted or replaced
+on-line.
+
+The most important concept of this document is that a managed node presents a
+well-known interface, executes according to a well-known life cycle state
+machine, and otherwise can be considered a black box. This allows freedom to
+the node developer on how they provide the managed life cycle functionality,
+while also ensuring that any tools created for managing nodes can work with any
+compliant node.
+
+
+## Life cycle
+
+![The proposed node life cycle state machine](/img/node_lifecycle/life_cycle_sm.png "The proposed node life cycle state machine")
+
+The behaviour of each state is as defined below.
+
+### Created state
+
+This is the life cycle state the node is in immediately after being
+instantiated. The node may transition to the `inactive` state by being
+initialised.
+
+Implementation note: In object-oriented languages, the onCreated method can be
+mapped to the class constructor.
+
+### Inactive state
+
+This state represents a node that is not currently performing any processing.
+While in this state, the node will not receive any execution time to read
+topics, perform processing of data, respond to service requests, etc. Any data
+that arrives on topics will not be read. Any service requests will not be
+answered.
+
+[To discuss: All of the above, and additionally should data/service requests be
+queued for processing when the node is made active?]
+
+Entering the `inactive` state from the `created` state requires running the
+node's `onInitialised` method. The node uses this to set up any resources it
+must hold throughout its life (irrespective of if it is active or inactive).
+Such resources may include topic publications and subscriptions, memory that is
+held continuously, and initialising configuration parameters.
+
+Entering the `inactive` state from the `active` state requires running the
+node's `onDeactivated` method. See the `active` state description for more
+information.
+
+Entering the `inactive` state from the `fatal error` state requires running
+the node's `onReset` method. See the `fatal error` state description for more
+information.
+
+If an error occurs while in this state that cannot be handled by the node, then
```

Comment:
As shown in the life cycle diagram, the inactive state may also transition to the fatal error state.

The reasoning is that the onActivate method may cause an error. The drawback of this transition is that going through the onError transition, it may not be possible to know whather the node was active when the error occurred or not. For this reason, I am open to alternatives to handling errors surrounding the inactive state, and also comments about whether the state transition history should be available to node developers.